fix: keep consistency with backend email validation

Antasel opened the pull request:
### Fixed Issues
$ https://github.com/Expensify/App/issues/17387
Proposal: https://github.com/Expensify/App/issues/17387#issuecomment-1514794314, https://github.com/Expensify/App/issues/17387#issuecomment-1630415664

# Tests

1. Go to a chat and send following valid emails and verify that all are parsed as email
<details>
  <summary>Valid emails</summary>

```
        // Domain length (63 chars in each label)
        test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com
        abc@abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890.km
        abc@co.abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890.km

        // Address length (64 chars)
        sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@test.com

        // Overall length (254 chars)
        averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com

        // Domain with lots of dashes
        sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asj-j-s-sjdjdjdjd-jdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdke.com.ab.net.aa.bb.cc.dd.ee
        abc@g---m--ai-l.com

        // Domain with repeated labels of 63 chars
        test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.com

        // TLD >=2 chars
        abc@gmail.co
        a@a.abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijk

        // Very short address
        a@example.com

        // xn-- style domain name
        test@xn--diseolatinoamericano-76b.com

        // Unusual but valid prefixes
        -test@example.com
        _test@example.com
        #test@example.com
        test.+123@example.com
        -test-@example.com
```
</details>

2. send following invalid emails and verify that the emails are not parsed as email, but only valid parts are extracted as email
<details>
  <summary>Invalid emails</summary>

```
        $test@gmail.com
        !test@gmail.com
        "test"@gmail.com
        ~abc@gmail.com~
        abc@gmail.com~
        test@example_123site.com
        test{@example.com
        test..new@example.com
        test@example-.a.com
        test@example......a.com

        // Invalid period location
        .test@example.com
        .test.new@example.com
        test.@example.com

        // Domain too long (>63 chars in each label)
        test@averylongdomainpartoftheemailthatwillgooverthelimitasitismorethan63chars.com
        abc@abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890a.km
        abc@co.abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890a.km

        // Address too long (>64 chars)
        averylongaddresspartoftheemailthatwillgovoerthelimitasitismorethan64chars@example.com

        // Incorrect domains start/end
        test@example-.com
        test@-example-.com

        // TLD too short
        test@example.a

        // TLD too long
        a@a.abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijkl
```
</details>

3. Navigate to Settings > Profile > Contact Method > New Contact Method, try an valid / invalid email

Demo Video:


https://github.com/Expensify/expensify-common/assets/139204153/79a254d7-1a54-4ecd-8a06-2d83840ae642





# Autotest cases:

<details>
  <summary>Str-test.js new tests</summary>

```
describe('Str.isValidEmail', () => {
    it('Correctly identifies valid emails', () => {
        expect(Str.isValidEmail('abc@gmail.com')).toBeTruthy();

        // Domain length (63 chars in each label)
        expect(Str.isValidEmail('test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com')).toBeTruthy();
        expect(Str.isValidEmail('abc@abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890.km')).toBeTruthy();
        expect(Str.isValidEmail('abc@co.abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890.km')).toBeTruthy();

        // Address length (64 chars)
        expect(Str.isValidEmail('sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@test.com')).toBeTruthy();

        // Overall length (254 chars)
        expect(Str.isValidEmail('averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com')).toBeTruthy();

        // Domain with lots of dashes
        expect(Str.isValidEmail('sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asj-j-s-sjdjdjdjd-jdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdke.com.ab.net.aa.bb.cc.dd.ee')).toBeTruthy();
        expect(Str.isValidEmail('abc@g---m--ai-l.com')).toBeTruthy();

        // Domain with repeated labels of 63 chars
        expect(Str.isValidEmail('test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.com')).toBeTruthy();

        // TLD >=2 chars
        expect(Str.isValidEmail('abc@gmail.co')).toBeTruthy();
        expect(Str.isValidEmail('a@a.abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijk')).toBeTruthy();

        // Very short address
        expect(Str.isValidEmail('a@example.com')).toBeTruthy();

        // xn-- style domain name
        expect(Str.isValidEmail('test@xn--diseolatinoamericano-76b.com')).toBeTruthy();

        // Unusual but valid prefixes
        expect(Str.isValidEmail('-test@example.com')).toBeTruthy();
        expect(Str.isValidEmail('_test@example.com')).toBeTruthy();
        expect(Str.isValidEmail('#test@example.com')).toBeTruthy();
        expect(Str.isValidEmail('test.+123@example.com')).toBeTruthy();
        expect(Str.isValidEmail('-test-@example.com')).toBeTruthy();

        // Invalid chars
        expect(Str.isValidEmail('$test@gmail.com')).toBeFalsy();
        expect(Str.isValidEmail('!test@gmail.com')).toBeFalsy();
        expect(Str.isValidEmail('"test"@gmail.com')).toBeFalsy();
        expect(Str.isValidEmail('~abc@gmail.com~')).toBeFalsy();
        expect(Str.isValidEmail('abc@gmail.com~')).toBeFalsy();
        expect(Str.isValidEmail('test@example_123site.com')).toBeFalsy();
        expect(Str.isValidEmail('test{@example.com')).toBeFalsy();
        expect(Str.isValidEmail('test..new@example.com')).toBeFalsy();
        expect(Str.isValidEmail('test@example-.a.com')).toBeFalsy();
        expect(Str.isValidEmail('test@example......a.com')).toBeFalsy();

        // Invalid period location
        expect(Str.isValidEmail('.test@example.com')).toBeFalsy();
        expect(Str.isValidEmail('.test.new@example.com')).toBeFalsy();
        expect(Str.isValidEmail('test.@example.com')).toBeFalsy();

        // Domain too long (>63 chars in each label)
        expect(Str.isValidEmail('test@averylongdomainpartoftheemailthatwillgooverthelimitasitismorethan63chars.com')).toBeFalsy();
        expect(Str.isValidEmail('abc@abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890a.km')).toBeFalsy();
        expect(Str.isValidEmail('abc@co.abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ01234567890a.km')).toBeFalsy();

        // Address too long (>64 chars)
        expect(Str.isValidEmail('averylongaddresspartoftheemailthatwillgovoerthelimitasitismorethan64chars@example.com')).toBeFalsy();

        // Incorrect domains start/end
        expect(Str.isValidEmail('test@example-.com')).toBeFalsy();
        expect(Str.isValidEmail('test@-example-.com')).toBeFalsy();

        // TLD too short
        expect(Str.isValidEmail('test@example.a')).toBeFalsy();

        // TLD too long
        expect(Str.isValidEmail('a@a.abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijkl')).toBeFalsy();
    });
});
```
</details>

<details>
  <summary>ExpensiMark-HTML-test.js new tests</summary>

```
// Check emails within other markdown
test('Test emails within other markdown', () => {
    const testString = '> test@example.com\n'
    + '```test@example.com```\n'
    + '`test@example.com`\n'
    + '_test@example.com_ '
    + '__test@example.com_';
    const result = '<blockquote><a href="mailto:test@example.com">test@example.com</a></blockquote>'
    + '<pre>test@example.com</pre>'
    + '<code>test@example.com</code><br />'
    + '<em><a href="mailto:test@example.com">test@example.com</a></em> '
    + '<em><a href="mailto:_test@example.com">_test@example.com</a></em>';
    expect(parser.replace(testString)).toBe(result);
});

// Check email regex's validity at various limits
test('Test markdown replacement for valid emails', () => {
    const testString = 'A simple email: abc@gmail.com, '
    + 'or a very short one a@example.com '
    + 'hitting the maximum domain length (63 chars) test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com '
    + 'or the maximum address length (64 chars) sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@test.com '
    + 'overall length of 254 averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com '
    + 'domain with many dashes sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asj-j-s-sjdjdjdjd-jdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdke.com.ab.net.aa.bb.cc.dd.ee '
    + ' how about a domain with repeated labels of 63 chars test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.com '
    + 'max length email with italics '
    + '_averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com_ '
    + ' xn-- style domain test@xn--diseolatinoamericano-76b.com '
    + 'or a more complex case where we need to determine where to apply italics markdown '
    + '_email@test.com\n_email2@test.com\n\nemail3@test.com_ '
    + 'some unusual, but valid prefixes -test@example.com '
    + ' and _test@example.com '
    + 'a max length email enclosed in brackets '
    + '(averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com) '
    + 'max length email with ellipsis ending '
    + 'averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com... '
    + 'try a markdown link with a valid max-length email '
    + '[text](sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com.a.aa.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjasfff)';
    const result = 'A simple email: <a href="mailto:abc@gmail.com">abc@gmail.com</a>, '
    + 'or a very short one <a href="mailto:a@example.com">a@example.com</a> '
    + 'hitting the maximum domain length (63 chars) <a href="mailto:test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com">test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com</a> '
    + 'or the maximum address length (64 chars) <a href="mailto:sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@test.com">sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@test.com</a> '
    + 'overall length of 254 <a href="mailto:averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com">averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com</a> '
    + 'domain with many dashes <a href="mailto:sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asj-j-s-sjdjdjdjd-jdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdke.com.ab.net.aa.bb.cc.dd.ee">sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asj-j-s-sjdjdjdjd-jdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdke.com.ab.net.aa.bb.cc.dd.ee</a> '
    + ' how about a domain with repeated labels of 63 chars <a href="mailto:test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.com">test@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekasgasgasgasgashfnfn.com</a> '
    + 'max length email with italics '
    + '<em><a href="mailto:averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com">averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com</a></em> '
    + ' xn-- style domain <a href="mailto:test@xn--diseolatinoamericano-76b.com">test@xn--diseolatinoamericano-76b.com</a> '
    + 'or a more complex case where we need to determine where to apply italics markdown '
    + '<em><a href="mailto:email@test.com">email@test.com</a><br />'
    + '<a href="mailto:_email2@test.com">_email2@test.com</a><br /><br />'
    + '<a href="mailto:email3@test.com">email3@test.com</a></em> '
    + 'some unusual, but valid prefixes <a href="mailto:-test@example.com">-test@example.com</a> '
    + ' and <a href="mailto:_test@example.com">_test@example.com</a> '
    + 'a max length email enclosed in brackets '
    + '(<a href="mailto:averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com">averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com</a>) '
    + 'max length email with ellipsis ending '
    + '<a href="mailto:averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com">averylongaddresspartthatalmostwillreachthelimitofcharsperaddress@nowwejustneedaverylongdomainpartthatwill.reachthetotallengthlimitforthewholeemailaddress.whichis254charsaccordingtothePHPvalidate-email-filter.extendingthetestlongeruntilwereachtheright.com</a>... '
    + 'try a markdown link with a valid max-length email '
    + '<a href="mailto:sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com.a.aa.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjasfff">text</a>';
    expect(parser.replace(testString)).toBe(result);
});

test('Test markdown replacement for invalid emails', () => {
    const testString = 'Replace the valid email part within a string '
    + '.test@example.com '
    + '$test@gmail.com '
    + 'test..new@example.com '
    + 'Some chars that are not allowed within emails '
    + 'test{@example.com '
    + 'domain length over limit test@averylongdomainpartoftheemailthatwillgooverthelimitasitismorethan63chars.com '
    + 'address over limit averylongaddresspartoftheemailthatwillgovoerthelimitasitismorethan64chars@example.com '
    + 'overall length too long '
    + 'sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com.a.aa.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjasfffa '
    + 'invalid domain start/end '
    + 'test@example-.com '
    + 'test@-example-.com '
    + 'test@example.a '
    + '[text](sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com.a.aa.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjasfffa)';
    const result = 'Replace the valid email part within a string '
    + '.<a href="mailto:test@example.com">test@example.com</a> '
    + '$<a href="mailto:test@gmail.com">test@gmail.com</a> '
    + 'test..<a href="mailto:new@example.com">new@example.com</a> '
    + 'Some chars that are not allowed within emails '
    + 'test{@example.com '
    + 'domain length over limit test@averylongdomainpartoftheemailthatwillgooverthelimitasitismorethan63chars.com '
    + 'address over limit averylongaddresspartoftheemailthatwillgovoerthelimitasitismorethan64chars@example.com '
    + 'overall length too long '
    + 'sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com.a.aa.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjasfffa '
    + 'invalid domain start/end '
    + 'test@example-.com '
    + 'test@-example-.com '
    + 'test@example.a '
    + '[text](sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com.a.aa.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjasfffa)';
    expect(parser.replace(testString)).toBe(result);
});
```
</details>

aldo-expensify commented:
Failing to add @aimane-chnaif as reviewer 🤷

Antasel commented:
@aimane-chnaif 
I've fixed lint and restored autoemail and quote rule order and some tweaks on Italic rule
according to the change, autotest cases are updated

jjcoffee commented:
@aimane-chnaif Can you clarify what you mean by the case of "_ prefixed email inside italic markdown"? Do you mean:
1. Email with leading underscore (`_test@example.com`) wrapped in italic markdown: `__test@example.com_`
2. Normal email wrapped in italic markdown: `_test@example.com_`

jjcoffee commented on __tests__/ExpensiMark-HTML-test.js:
```diff
@@ -102,10 +102,10 @@ test('Test markdown replacement for emails wrapped in bold/strikethrough/italic
 
     testInput = '_abc@gmail.com_';
-    expect(parser.replace(testInput)).toBe('<em><a href="mailto:abc@gmail.com">abc@gmail.com</a></em>');
+    expect(parser.replace(testInput)).toBe('<a href="mailto:_abc@gmail.com">_abc@gmail.com</a>_');
```
I don't think this would be expected behaviour and would result in a regression. If you wrap a normal email (`abc@gmail.com`) in italic markdown it should just be italicised and linked. cc @aimane-chnaif

Antasel replied:
> @Antasel Let's add all of these in test case. Github markdown seems to have different logic
> 
> * `abc@gmail.com` -> [abc@gmail.com](mailto:abc@gmail.com)
> * `_abc@gmail.com` -> [_abc@gmail.com](mailto:_abc@gmail.com)
> * `__abc@gmail.com` -> [__abc@gmail.com](mailto:__abc@gmail.com)
> * `abc@gmail.com_` -> abc@gmail.com_
> * `abc@gmail.com__` -> abc@gmail.com__
> * `_abc@gmail.com_` -> _[abc@gmail.com](mailto:abc@gmail.com)_
> * `__abc@gmail.com_` -> __[abc@gmail.com](mailto:abc@gmail.com)_
> * `_abc@gmail.com__` -> _[abc@gmail.com](mailto:abc@gmail.com)__
> * `__abc@gmail.com__` -> **[abc@gmail.com](mailto:abc@gmail.com)**

those cases are already covered in https://github.com/Expensify/App/issues/17387#issuecomment-1630415664
so there is no need to restore, just leave it as it is. 
How you think ? @jjcoffee

Antasel replied:
@aimane-chnaif you meant "the case of _ prefixed email inside italic markdown can be skipped"  is following case ? 

`__abc@gmail.com_` -> __abc@gmail.com_
after parsed, underscore is not contained in email
if so, it can be done by only moving autoemail rule after italic rule. no need extra tweak in italic rule
cc: @aimane-chnaif

jjcoffee replied:
@Antasel Sorry I'm not clear, what would Do you mean the result for `__abc@gmail.com_` will be

jjcoffee commented on __tests__/ExpensiMark-HTML-test.js:
```diff
@@ -102,10 +102,10 @@ test('Test markdown replacement for emails wrapped in bold/strikethrough/italic
 
     testInput = '~*_abc@gmail.com_*~';
-    expect(parser.replace(testInput)).toBe('<del><strong><em><a href="mailto:abc@gmail.com">abc@gmail.com</a></em></strong></del>');
+    expect(parser.replace(testInput)).toBe('<del><strong><a href="mailto:_abc@gmail.com">_abc@gmail.com</a>_</strong></del>');
```
Same here, I don't think this is the expected behaviour.

jjcoffee commented:
@Antasel Let's add some of the markdown tests from `ExpensiMark-HTML-test.js` into the **Tests** section of the PR so it's easier to visualise how it will work for markdown cases. For example `_abc@gmail.com_` would be useful, maybe some of the multiline ones too?

jjcoffee commented:
@Antasel Please can you fix the lint issues again.

jjcoffee commented:
@Antasel See [here](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#begin-coding-your-solution-in-a-pull-request) at point 10.  It might be easiest to close this PR and open a new one where you make new signed commits, then you'll get a "Verified" badge here:
![image](https://github.com/Expensify/expensify-common/assets/27287420/f527c845-ac07-404c-8510-f224d1bbcd2d)

Antasel commented:
thanks, I will do it now,
but I remember I followed it before

jjcoffee commented:
The result is not correct for `.test.new@example.com`:

![image](https://github.com/Expensify/expensify-common/assets/27287420/365dfd79-558f-40eb-9a8f-4006e168f82d)

Also doesn't work with it on its own. Can you check @Antasel?

jjcoffee commented:
@aimane-chnaif Are you saying it's incorrect to parse `test.new@example.com` as an email, or just the one with the leading dot? The former is correct as far as I know!

Antasel commented:
![Screenshot_1](https://github.com/Expensify/expensify-common/assets/139204153/dc0a0256-cd70-40ed-8bfb-db3a3632a94f)
@jjcoffee I am getting correctly

Antasel commented:
yes sure

Antasel commented:
@jjcoffee which platform did you check it on ?

jjcoffee commented:
@Antasel I think _all_ your commits need to be signed - you've only signed the new one. I think that's correct right @aimane-chnaif?

I will retest soon - I might have some cached node_modules messing things up :smile:

jjcoffee commented:
@Antasel It fails when I add it as a test, so it must be an issue with the regex:

![image](https://github.com/Expensify/expensify-common/assets/27287420/fcb00770-f4cc-488d-9bec-ead1679442a2)

jjcoffee commented:
@Antasel Is that passing the test now?

jjcoffee commented:
@aimane-chnaif It's not exactly the same as BE also applies a secondary regex (`[\w.'#%+-]+@[a-zA-Z\d.-]+.[a-zA-Z]{2,}`), or do you mean apply the exact PHP regex, then apply the secondary regex? I'm unsure if that might have a performance impact.

Antasel commented:
![Screenshot_2](https://github.com/Expensify/expensify-common/assets/139204153/dad7c1dc-1e33-4d18-bf9d-3f75285551b4)

@jjcoffee all passed.
@aimane-chnaif it's ready for review

jjcoffee commented:
@Antasel Still failing for me, can you add the relevant test? e.g.
```
testInput = 'test.new@gmail.com';
expect(parser.replace(testInput)).toBe('<a href="mailto:test.new@gmail.com">test.new@gmail.com</a>');
```

Antasel commented:
@jjcoffee of course, it's passed
did you copy & paste the regex what i mentioned ?

Antasel commented:
I tried same test case, but failed
it's because I moved autoemail rule after autolink rule. so autolink rule is applied first, 'new' is in the predefined TLD list.
so test.new was parsed as url link
to prevent it, i added `\@(?:[a-z\d-]+\.)+[a-z]{2,})` in negative lookbehind of autolink regex
```
`(?![^<]*>|[^<>]*<\\/)([_*~]*?)${MARKDOWN_URL_REGEX}\\1(?!(@(?:[a-z\\d-]+\\.)+[a-z]{2,})|((?:(?!<a).)+)?<\\/a>|[^<]*(<\\/pre>|<\\/code>))`,
```
then issue is fixed
how do you think about the tweak ?

Antasel commented:
i think it rather can help to fix https://github.com/Expensify/App/issues/16762
the issue is produced since of first parsed email, so url is not parsed properly.
`https://staging.new.expensify.com/details?login=testing@gmail.com`
here, if autoemail is before autolink, testing@gmail.com is parsed firstly, then entire url is not parsed correctly.

Antasel commented:
I think current approach has fixed the regression issue

Antasel commented:
@aimane-chnaif I posted a solution https://github.com/Expensify/App/issues/16762#issuecomment-1682687000, which is covering well on the basis of current approach.
all test cases are passed there

Antasel commented:
I think this PR is ready for review
@aimane-chnaif @jjcoffee

jjcoffee commented on lib/ExpensiMark.js:
```diff
@@ -142,7 +128,7 @@ export default class ExpensiMark {
                 process: (textToProcess, replacement) => {
                     const regex = new RegExp(
-                        `(?![^<]*>|[^<>]*<\\/)([_*~]*?)${MARKDOWN_URL_REGEX}\\1(?!((?:(?!<a).)+)?<\\/a>|[^<]*(<\\/pre>|<\\/code>))`,
+                        `(?![^<]*>|[^<>]*<\\/)([_*~]*?)${MARKDOWN_URL_REGEX}\\1(?!(@(?:[a-z\\d-]+\\.)+[a-z]{2,})|((?:(?!<a).)+)?<\\/a>|[^<]*(<\\/pre>|<\\/code>))`,
```
Wouldn't it be better to add the negative lookahead directly in `URL_WEBSITE_REGEX` so that it applies before the `URL_PARAM_REGEX` (like was proposed [here](https://github.com/Expensify/App/issues/16762#issuecomment-1500807823))?

jjcoffee commented:
@Antasel I've added a couple comments.

Antasel commented:
@jjcoffee done.

I've also removed space and \\b on autoemail regex, would be `(?:(?<=[^\\w'#%+-])|^)` instead of `(?:(?<= |[^\\w'#%+-])|^|\\b)`

Antasel commented:
@aimane-chnaif @jjcoffee 
I've found a issue that lookbehind (?<=) is not supported on Safari version less than 16.4 (https://caniuse.com/js-regexp-lookbehind)

so as a alternative, it can be as following: 
```
         {
                name: 'autoEmail',
                regex: new RegExp(
                    `(?![^<]*>|[^<>]*<\\/(?!em))([^\\w'#%+-]|^|<em>)${CONST.REG_EXP.MARKDOWN_EMAIL}(?!((?:(?!<a).)+)?<\\/a>|[^<]*(<\\/pre>|<\\/code>))`,
                    'gim',
                ),
                replacement: '$1<a href="mailto:$2">$2</a>',
            },
```

Antasel commented:
> @Antasel Good catch with the lookbehind not being supported! Could you explain the reasoning behind the recent changes:
> 
> 1. `<em>` added to `([^\\w'#%+-]|^|<em>)`
> 2. [Removed](https://github.com/Expensify/expensify-common/pull/566#issuecomment-1685802959) space and \b on autoemail regex

1. autoemail rule is moved after italic rule, so <em> is added before parsing autoemail. so need to add
2. (?<=[^\\w'#%+-]) was also covering (?<= ), and \\b

jjcoffee commented:
@Antasel Thanks for the explanation! So do we still need `(?!em)` in the negative lookahead `(?![^<]*>|[^<>]*<\\/(?!em))`?

Also can you ensure that all cases that `(?![^<]*>)` would exclude are covered, e.g. I think <h1> won't match currently.

Antasel commented:
> So do we still need `(?!em)` in the negative lookahead `(?![^<]*>|[^<>]*<\\/(?!em))`?

yes



> Also can you ensure that all cases that `(?![^<]*>)` would exclude are covered, e.g. I think `<h1>` won't match currently.

@jjcoffee I agree on your case, so need to move heading1 rule after autoemail

Antasel commented:
@jjcoffee we have other issue on ios.
a err 'Maximum Regex stack depth reached' is occurred when send following invalid email

jjcoffee commented:
@Antasel Oh no! I wonder if it's due to the email regex being overly complex. Can you try removing all the length params from the regex and see if you get the same error? If that helps, we can then just switch to checking the captured email's length with either a simple JS function that checks each part of the email or just apply a separate regex that only checks the lengths.

Antasel commented:
no, it's not related to new email regex.
I tested with original expensify-common version, just after setting domain label length limitation on markdown email regex of autoemail. so the invalid email has not been parsed by autoemail.
and when it's parsed by autolink right after it, same error is occurred

we can see that it's old issue, not caused by new approach of this PR.
just it was not checked because most invalid email was parsed as email, then skipped by autolink

jjcoffee commented:
@Antasel I'm confused, are you saying that autolink is trying to parse the domain part of this email:

`sjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcjab@asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.com.a.aa.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjdkekejdcjdkeekcj.asjjssjdjdjdjdjdjjeiwiwiwowkdjdjdieikdjfidekjcjasfffa`

jjcoffee commented:
@Antasel Can you try merging main in case the issue is resolved by [the PR](https://github.com/Expensify/expensify-common/pull/559) @aimane-chnaif mentions?

Antasel commented:
noted, let me try

Antasel commented:
> @Antasel I'm confused, are you saying that autolink is trying to parse the domain part of this email:

yes, sure
if we disable autolink rule, no error is happened

Antasel commented:
@jjcoffee @aimane-chnaif 
you can try at your side.
on latest version of expensify-common, please limit domain level length to 10, then autolink rule is not parsing even simple invalid email `T@aaaaaaaaaaaa.aaaaaaaaaaaa.aaaaa` , causing the error.
maybe  https://github.com/Expensify/expensify-common/pull/559 's contributor should check it again

Antasel commented:
@aimane-chnaif @jjcoffee 
I've posted more reasonable bug report (https://github.com/Expensify/App/issues/21266#issuecomment-1691111696) on [#21266](https://github.com/Expensify/App/issues/21266) .
The issue is out of current PR's scope.
I think we can move forward.

jjcoffee commented:
@Antasel Agree it looks like a separate issue. From what I understand, any invalid email (like a super long email that's over the length limits) will not be parsed by the email rule so the autolinker tries to parse it and fails because of issues with the autolink rule. I think it's worth reporting this as a separate bug on Slack. 

I think next steps here are to add some additional manual tests:

1. Try to login with an invalid email + an email at the limit (e.g. just below length limit)
2. Navigate to Settings > Profile > Contact Method > New Contact Method, try an invalid email

I will run some platform tests today, but otherwise it's ready for review. We're running close to the 9 day limit for merging, so let's try and get this merged today! cc @aimane-chnaif

Antasel commented:
@jjcoffee noted
Could you please make video of iOS, mobile safari, desktop Platform to save time, whilst testing ?

jjcoffee commented:
@Antasel Actually we can ignore tests for sign-in as if FE validates, BE will still return that it's an invalid email (the error message is just slightly different). Also, there's now(?) a length limit on the email input field, so you can't enter an email >254 chars anyway :smile: 

Can you add `Navigate to Settings > Profile > Contact Method > New Contact Method, try an invalid email` as a test to the description here please?

Antasel commented:
> Can you add `Navigate to Settings > Profile > Contact Method > New Contact Method, try an invalid email` as a test to the description here please? Maybe specifically for `a@a.abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijkl`

I've added it and updated demo video.
now Could we move forward ?
@aimane-chnaif @jjcoffee

jjcoffee commented:
@Antasel Thanks! Please also remove the `// Overall length too long` item in the invalid emails list since this will always fail the test on iOS.

jjcoffee commented:
Tests well, just waiting for the Android build. I think this is ready for review @aimane-chnaif - please prioritise as we're getting close to the 9 day limit. Thanks!

### Screenshots/Videos
<details>
<summary>Web</summary>

**New Contact method**

https://github.com/Expensify/expensify-common/assets/27287420/0294ddc2-c527-42cf-a6d2-fd2fdb09a0b3

**Markdown**

https://github.com/Expensify/expensify-common/assets/27287420/03ecb58f-c05f-4014-9a95-6fe1271b0565

</details>

<details>
<summary>Mobile Web - Chrome</summary>

**New Contact method**

https://github.com/Expensify/expensify-common/assets/27287420/b7fde9dc-31ff-46bd-969d-85a69284f34f

**Markdown - valid**

https://github.com/Expensify/expensify-common/assets/27287420/214f067b-b6b7-4716-b4e3-8cbc523504d8

**Markdown - invalid**

Uploading android-chrome-markdown-invalid.mp4…

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/expensify-common/assets/27287420/086cd72b-a021-4beb-97b0-d6f380b5ad45

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/expensify-common/assets/27287420/d3dab962-d524-4fba-9e5a-dfbc8bc45fa7

</details>

<details>
<summary>iOS</summary>

**New Contact method**

https://github.com/Expensify/expensify-common/assets/27287420/44854210-16e2-4d83-85a3-fadde57b4471

**Markdown**

https://github.com/Expensify/expensify-common/assets/27287420/6bb3c200-02a7-49f5-bbe3-4c2dc015c4c8

</details>

<details>
<summary>Android</summary>


</details>

Antasel commented:
> @Antasel Thanks! Please also remove the `// Overall length too long` item in the invalid emails list since this will always fail the test on iOS.

It's removed

jjcoffee commented:
Android test added! @aimane-chnaif Do you think you can review today?

jjcoffee commented:
I guess this is to do with putting the `autoEmail` rule after the `italic` rule, since `a~*_b@gmail.com` would otherwise be parsed by `autoEmail` as this being the email part: `_b@gmail.com`.

jjcoffee commented:
@aimane-chnaif Actually this is current behaviour on main, so unrelated to this PR:
![image](https://github.com/Expensify/expensify-common/assets/27287420/621e57ac-b858-4cfd-b1cb-3731941d5337)

NikkiWines commented on __tests__/Str-test.js:
```diff
@@ -129,10 +120,80 @@ describe('Str.fromCurrencyToNumber', () => {
-    it('Correctly detects a valid email', () => {
+    it('Correctly identifies valid emails', () => {
         expect(Str.isValidEmail('abc@gmail.com')).toBeTruthy();
-        expect(Str.isValidEmail('test@gmail')).toBeFalsy();
-        expect(Str.isValidEmail('@gmail.com')).toBeFalsy();
-        expect(Str.isValidEmail('usernamelongerthan64charactersshouldnotworkaccordingtorfc822whichisusedbyphp@gmail.com')).toBeFalsy();
```
All of these should still be invalid, right? Why not retain these tests?

NikkiWines replied:
Though looks like at least the last one is already covered by the test on 184

Antasel replied:
@NikkiWines @aimane-chnaif 
The cases have been restored.

NikkiWines reviewed (commented):
Amazing 🤩 one very minor question but otherwise looks good.

s77rt commented on lib/ExpensiMark.js:
```diff
@@ -153,6 +139,45 @@ export default class ExpensiMark {
                     return `${g1}<a href="${href}" target="_blank" rel="noreferrer noopener">${g2}</a>${g1}`;
                 },
             },
+            {
+                /**
+                 * Use \b in this case because it will match on words, letters,
+                 * and _: https://www.rexegg.com/regex-boundaries.html#wordboundary
+                 * The !_blank is to prevent the `target="_blank">` section of the
+                 * link replacement from being captured Additionally, something like
+                 * `\b\_([^<>]*?)\_\b` doesn't work because it won't replace
+                 * `_https://www.test.com_`
+                 * Use [\s\S]* instead of .* to match newline
+                 */
+                name: 'italic',
+                regex: /(\b_+|\b)(?!_blank")_((?![\s_])[\s\S]*?[^\s_])_(?![^\W_])(?![^<]*(<\/pre>|<\/code>|<\/a>|<\/mention-user>|_blank))/g,
+
+                // We want to add extraLeadingUnderscores back before the <em> tag unless textWithinUnderscores starts with valid email
+                replacement: (match, extraLeadingUnderscores, textWithinUnderscores) => {
+                    if (textWithinUnderscores.includes('<pre>') || this.containsNonPairTag(textWithinUnderscores)) {
+                        return match;
+                    }
+                    if (String(textWithinUnderscores).match(`^${CONST.REG_EXP.MARKDOWN_EMAIL}`)) {
+                        return `<em>${extraLeadingUnderscores}${textWithinUnderscores}</em>`;
+                    }
+                    return `${extraLeadingUnderscores}<em>${textWithinUnderscores}</em>`;
+                },
+            },
```
We used to have the quote rule run before the italic rule. Changing that order caused an invalid markdown rendering for cases like:

```
_Msg then a
> quote_
```

Coming from https://github.com/Expensify/App/issues/26941
